pkg/resource/internal/kv/kv.go

```diff
@@ -0,0 +1,161 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package kv implements core key-value type as a base for labels and annotations.
+package kv
+
+import (
+	"sort"
+
+	"gopkg.in/yaml.v3"
+)
+
+// KV is a set free-form of key-value pairs.
+//
+// Order of keys is not guaranteed.
+//
+// KV support copy-on-write semantics, so metadata copies share common labels as long as possible.
+type KV struct {
+	m map[string]string
+}
+
+// Delete the key.
+//
+// Deleting the key copies the map, so metadata copies share common storage as long as possible.
+func (kv *KV) Delete(key string) {
+	if _, ok := kv.m[key]; !ok {
+		// no change
+		return
+	}
+
+	kvCopy := make(map[string]string, len(kv.m))
+
+	for k, v := range kv.m {
+		kvCopy[k] = v
```

Comment:
is it easier to check if the key to delete matches here and just continue?

Reply:
fixed!